drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,572 @@ impl ServerImpl {
         // packed against the end of the `Message`.  Let's shift it backwards
         // based on the size of the leading `Message`:
         tx_data_buf.copy_within(
-            Message::MAX_SIZE..(Message::MAX_SIZE + data_len),
-            msg_len,
+            reserved_framing..(reserved_framing + data_len),
+            hdr_len + msg_len,
         );
-        Some((msg_len + data_len) as u32)
+        Some((hdr_len + msg_len + data_len) as u32)
     }
 
     fn handle_host_request(
         &mut self,
         h: HostRequest,
-        modules: ModuleId,
         data: &[u8],
         out: &mut [u8],
-    ) -> Result<(SpResponse, usize), Error> {
+    ) -> (MessageBody, usize) {
         match h {
-            HostRequest::Status => {
+            HostRequest::Status(modules) => {
                 ringbuf_entry!(Trace::Status(modules));
-                let count = self.get_status(modules, out)?;
-                Ok((SpResponse::Status, count))
+                let mask = LogicalPortMask(modules.0 as u32);
+                let (num_status_bytes, result) = self.get_status(mask, out);
+                ringbuf_entry!(Trace::OperationResult(result));
+                let success = ModuleId(result.success().0 as u64);
+                let (final_payload_len, failed_modules) = self.handle_errors(
+                    modules,
+                    result,
+                    None,
+                    num_status_bytes,
+                    out,
+                );
+
+                (
+                    MessageBody::SpResponse(SpResponse::Status {
+                        modules: success,
+                        failed_modules,
+                    }),
+                    final_payload_len,
+                )
             }
-            HostRequest::Read(mem) => {
-                ringbuf_entry!(Trace::Read(modules, mem));
-                let out_size = mem.len() as u32 * modules.ports.0.count_ones();
-                if out_size as usize > transceiver_messages::MAX_PAYLOAD_SIZE {
-                    return Err(Error::RequestTooLarge);
+            HostRequest::Read { modules, read } => {
+                ringbuf_entry!(Trace::Read(modules, read));
+                let mask = LogicalPortMask(modules.0 as u32);
+                let request_size = read.len() as u32 * mask.0.count_ones();
+                if request_size as usize
+                    > transceiver_messages::MAX_PAYLOAD_SIZE
+                {
+                    return (
+                        MessageBody::Error(ProtocolError::RequestTooLarge),
+                        0,
+                    );
                 }
-                self.read(mem, modules, out)?;
-                Ok((SpResponse::Read(mem), out_size as usize))
+
+                let result = self.read(read, mask, out);
+                ringbuf_entry!(Trace::OperationResult(result));
+                let success = ModuleId(result.success().0 as u64);
```

Comment:
The other direction of conversions might be useful too? `impl From<LogicalPortMask> for ModuleId`?